llvm/lib/Target/AMDGPU/SIInstructions.td

```diff
@@ -1131,25 +1131,42 @@ multiclass f16_fp_Pats<Instruction cvt_f16_f32_inst_e64, Instruction cvt_f32_f16
     (cvt_f32_f16_inst_e64 SRCMODS.NEG, $src0)
   >;
 
+  // fp_to_fp16 patterns
   def : GCNPat <
-    (f64 (any_fpextend f16:$src)),
-    (V_CVT_F64_F32_e32 (cvt_f32_f16_inst_e64 SRCMODS.NONE, $src))
+    (i32 (AMDGPUfp_to_f16 (f32 (VOP3Mods f32:$src0, i32:$src0_modifiers)))),
+    (cvt_f16_f32_inst_e64 $src0_modifiers, f32:$src0)
   >;
 
-  // fp_to_fp16 patterns
+  // This is only used on targets without half support
+  // TODO: Introduce strict variant of AMDGPUfp_to_f16 and share custom lowering
   def : GCNPat <
-    (i32 (AMDGPUfp_to_f16 (f32 (VOP3Mods f32:$src0, i32:$src0_modifiers)))),
+    (i32 (strict_fp_to_f16 (f32 (VOP3Mods f32:$src0, i32:$src0_modifiers)))),
     (cvt_f16_f32_inst_e64 $src0_modifiers, f32:$src0)
   >;
+}
+
+let SubtargetPredicate = NotHasTrue16BitInsts in
+defm : f16_to_fp_Pats<V_CVT_F16_F32_e64, V_CVT_F32_F16_e64>;
+
+let SubtargetPredicate = UseFakeTrue16Insts in
+defm : f16_to_fp_Pats<V_CVT_F16_F32_fake16_e64, V_CVT_F32_F16_fake16_e64>;
+
+multiclass f16_fp_Pats<Instruction cvt_f16_f32_inst_e64,
+                       Instruction cvt_f32_f16_inst_e64,
+                       RegOrImmOperand VSrc> {
+  def : GCNPat <
+    (f64 (any_fpextend f16:$src)),
+    (V_CVT_F64_F32_e32 (cvt_f32_f16_inst_e64 SRCMODS.NONE, $src))
+  >;
```

Comment:
This is wrong, you can't do this conversion in 2 steps. f64->f16 needs to go through custom lowering

Reply:
Hi Matt. Can you help to elaborate this a bit?

I am actually moving this pattern into another location instead of adding it. It's originally here https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/AMDGPU/SIInstructions.td#L1135

Reply:
That one should also be fixed / removed. f64->f16 is one rounding. f64->f32->f16 is 2 rounding steps. You would need fast math to do this. This pattern might just be dead anyway, we have a custom expansion for it

Reply:
I see. Let me do this change in another PR

Reply:
Hi Matt. Can you share any link to the custom expansion for the f64->f16 rounding? I am trying to see how can I address them in the test. Thanks!

Reply:
https://github.com/llvm/llvm-project/blob/ee08d9cba5615937acf28087da841886cc6a0144/llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp#L3518

We ended up lowering it through FP_TO_FP16, presumably for the benefit of the targets without legal f16

Reply:
Hi Matt. Thanks for the link!

I took a quick look at this. If I understand correctly this pattern seems to be a f16 to f64 fp extension and it seems ok to me.  I check the other tablegen files but I didn't find any pattern that converting f64 to f16, seems we don't have any test for f64 to f16 as well, not sure if I missed it

Reply:
From these comments it sound like this pattern should be removed
```
def : GCNPat <
    (f64 (any_fpextend f16:$src)),
    (V_CVT_F64_F32_e32 (cvt_f32_f16_inst_e64 SRCMODS.NONE, $src))
  >;
```

Reply:
I think this was a misunderstanding? This should be a f16->f64 instead of f64->f16 so it should be fine

Reply:
Oh I see, thanks for clarifying.